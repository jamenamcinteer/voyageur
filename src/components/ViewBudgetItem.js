import React from "react";
import { Link } from "react-router-dom";
import TripHeader from "./Navigation/TripHeader";
import { Container } from "./StyledComponents/Layout";
import ButtonLink from "./Buttons/ButtonLink";
import styled from "styled-components";
import useSortDescendingNumerical from "../hooks/useSortDescendingNumerical";
import moment from "moment";
import { connect } from "react-redux";

const BudgetItemHeader = styled.h3`
  width: 100%;
  background: ${props => props.theme.themeColorSecondaryLight};
  font-size: 1.2em;
  font-weight: normal;
  text-align: center;
  padding: 7px 0;
  margin-bottom: 0;
`;

const BudgetItemNotes = styled.p`
  border-bottom: 1px solid ${props => props.theme.themeColorSecondary};
  margin: 0;
  padding: 5px 20px;
  color: ${props => props.theme.lightFont};
  font-size: 1em;
  font-style: italic;
`;

const ExpenseItem = styled(Link)`
  border-bottom: 1px solid ${props => props.theme.themeColorSecondary};
  display: block;
  padding: 10px 20px;
  font-size: 1.2em;
  font-weight: normal;
  text-decoration: none;
  color: ${props => props.theme.darkFont};
`;

const ExpenseItemFirstLine = styled.div`
  display: flex;
  justify-content: space-between;
  padding-bottom: 5px;
`;

const ExpenseItemSecondLine = styled.div`
  display: flex;
  justify-content: space-between;
  padding-bottom: 5px;
`;

const ExpenseItemDate = styled.div`
  color: ${props => props.theme.darkFont};
  font-size: 0.8em;
  font-weight: normal;
`;

const ExpenseItemEdit = styled.div`
  color: ${props => props.theme.themeColor};
  font-size: 0.8em;
  font-weight: normal;
  text-decoration: underline;
`;

const ExpenseItemSummary = styled.h4`
  font-size: 1em;
  font-weight: normal;
  margin: 0;
`;

const ExpenseItemCost = styled.div``;

const ExpenseItemNotes = styled.p`
  color: ${props => props.theme.lightFont};
  font-size: 0.8em;
  font-weight: normal;
  font-style: italic;
  margin: 0;
`;

const ViewBudgetItem = props => {
  const trip = props.trips.find(trip => trip._id === props.match.params.id);

  // const budgetCategory = state.budgetCategories.find(
  //   budgetCategory => budgetCategory.id === props.match.params.budgetCategoryId
  // );

  const budgetItem = props.budgetItems.find(
    budgetItem => budgetItem._id === props.match.params.budgetItemId
  );

  const expenses = useSortDescendingNumerical(
    props.expenses.filter(
      i => i.budgetItemId === props.match.params.budgetItemId
    ),
    "date"
  );

  return (
    <div>
      {budgetItem && (
        <div>
          <TripHeader
            title="Expenses"
            theme={props.theme}
            backTo={`/trip/${trip._id}`}
            trip={trip}
            budgetItems={props.budgetItems}
            expenses={props.expenses}
            auth={props.auth}
          />
          <Container style={{ ...{ textAlign: "center" } }}>
            <ButtonLink
              to={`/trip/${trip._id}/budget-category/${
                props.match.params.budgetCategoryId
              }/budget-item/${budgetItem._id}/edit`}
              theme={props.theme}
              buttonType="secondary"
              buttonText="Edit Budget Item"
              customStyles={{
                background: {
                  display: "inline-block",
                  width: "auto"
                }
              }}
            />
            <ButtonLink
              to={`/trip/${trip._id}/add-expense?budgetCategory=${
                props.match.params.budgetCategoryId
              }&budgetItem=${budgetItem._id}`}
              buttonText="Add Expense"
              buttonType="primary"
              theme={props.theme}
              customStyles={{
                background: {
                  display: "inline-block",
                  width: "auto"
                }
              }}
            />
          </Container>
          <BudgetItemHeader>{budgetItem.budgetItem}</BudgetItemHeader>
          <BudgetItemNotes>{budgetItem.notes}</BudgetItemNotes>
          {expenses &&
            expenses.length > 0 &&
            expenses.map(expense => {
              return (
                <ExpenseItem
                  key={expense._id}
                  to={`/trip/${expense.tripId}/budget-category/${
                    expense.budgetCategoryId
                  }/budget-item/${expense.budgetItemId}/expense/${expense._id}`}
                >
                  <ExpenseItemFirstLine>
                    <ExpenseItemDate>
                      {moment(expense.date).format("MMM D")}
                    </ExpenseItemDate>
                    <ExpenseItemEdit>Edit</ExpenseItemEdit>
                  </ExpenseItemFirstLine>
                  <ExpenseItemSecondLine>
                    <ExpenseItemSummary>{expense.summary}</ExpenseItemSummary>
                    <ExpenseItemCost>${expense.cost}</ExpenseItemCost>
                  </ExpenseItemSecondLine>
                  <ExpenseItemNotes>{expense.notes}</ExpenseItemNotes>
                </ExpenseItem>
              );
            })}
        </div>
      )}
    </div>
  );
};

const mapStateToProps = state => {
  return {
    trips: state.trips,
    budgetItems: state.budgetItems,
    expenses: state.expenses,
    auth: state.auth
  };
};

export default connect(mapStateToProps)(ViewBudgetItem);
