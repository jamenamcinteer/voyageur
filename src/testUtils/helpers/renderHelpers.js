import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../../theme";
import "jest-styled-components";
import { createStore } from "redux";
import { Provider } from "react-redux";
import tripsReducer, { tripsReducerDefaultState } from "../../reducers/trips";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

export function renderWithTheme(component) {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
}

export function renderWithReduxRouterAndTheme(
  ui,
  {
    initialState,
    store = createStore(tripsReducer, tripsReducerDefaultState),
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) {
  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <Router history={history}>{ui}</Router>
        </ThemeProvider>
      </Provider>
    ),
    store,
    history
  };
}
