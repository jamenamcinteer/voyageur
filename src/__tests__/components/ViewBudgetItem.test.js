import React from "react";
import { cleanup } from "@testing-library/react";
import { createStore } from "redux";
import ViewBudgetItem from "../../components/ViewBudgetItem";
import {
  trips,
  budgetCategories,
  budgetItems,
  expenses,
  auth
} from "../../testUtils/fixtures/mockStoreData";
import { renderWithReduxRouterAndTheme } from "../../testUtils/helpers/renderHelpers";

afterEach(cleanup);

const store = createStore(() => ({
  trips,
  budgetCategories,
  budgetItems,
  expenses,
  auth
}));

test("can render with redux with custom store", () => {
  const { asFragment } = renderWithReduxRouterAndTheme(
    <ViewBudgetItem match={{ params: { id: "1", budgetItemId: "1" } }} />,
    {
      store
    }
  );
  expect(asFragment()).toMatchSnapshot();
});
