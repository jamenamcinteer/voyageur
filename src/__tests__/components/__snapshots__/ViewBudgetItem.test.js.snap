// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render with redux with custom store 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        <div>
          <div
            class="sc-bdVaJa hZqfbl"
          >
            <a
              href="/trip/1"
            >
              <svg
                aria-label="Go Back"
                style="fill: #4b5c6b; width: 20px; height: 20px; float: left; margin-top: 2px;"
                version="1.1"
                viewBox="0 0 199.404 199.404"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <polygon
                    points="135.412,0 35.709,99.702 135.412,199.404 163.695,171.119 92.277,99.702 163.695,28.285 	"
                  />
                </g>
              </svg>
            </a>
            <h1
              class="sc-bwzfXH OVfcq"
            >
              Expenses
            </h1>
            <button
              class="sc-htpNat fqCEYg"
            >
              <img
                alt="Account Options"
                class="sc-bxivhb jwtarI"
                src="https://picsum.photos/id/947/45/45"
              />
            </button>
          </div>
          <div
            style="position: relative;"
          >
            <div
              style="background-image: url(https://images.unsplash.com/photo-1549317487-92cec5e76645?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjc5ODc0fQ&w=300&h=100&fit=crop&crop=focalpoint); background-repeat: no-repeat; background-size: cover; filter: brightness(0.6); height: 150px;"
            />
            <div
              style="height: 150px; padding: 0px 20px; color: rgb(255, 255, 255); font-size: 1em; font-weight: normal; position: absolute; width: 100%; top: 0px;"
            >
              <h2
                class="sc-bZQynM krdFRo"
              >
                Hawaii
              </h2>
              <div
                style="display: flex; justify-content: space-between; align-items: center; padding-bottom: 10px;"
              >
                <h3
                  class="sc-gzVnrw HgWFV"
                >
                  Apr 26, 1970 - Oct 1, 8307
                </h3>
                <h3
                  class="sc-htoDjs btYDzX"
                >
                  $595 / $2000
                </h3>
              </div>
              <div
                style="width: 100%; height: 10px; position: relative; background: rgb(220, 224, 227); border-radius: 5px;"
              >
                <span
                  style="display: block; height: 100%; border-radius: 5px; position: relative; overflow: hidden; width: 29.731%; background-color: rgb(12, 124, 136);"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-dnqmqq eibGdB"
        style="text-align: center;"
      >
        <a
          backgroundcolor="#fff"
          border="1px solid undefined"
          buttontype="secondary"
          class="sc-gZMcBi iEljFg"
          href="/trip/1/budget-category/undefined/budget-item/1/edit"
          style="display: inline-block; width: auto;"
        >
          <span
            class="sc-gqjmRU enArve"
          >
            Edit Budget Item
          </span>
        </a>
        <a
          border="0"
          buttontype="primary"
          class="sc-gZMcBi ekpTYP"
          href="/trip/1/add-expense?budgetCategory=undefined&budgetItem=1"
          style="display: inline-block; width: auto;"
        >
          <span
            class="sc-gqjmRU ihIhkB"
            color="#fff"
          >
            Add Expense
          </span>
        </a>
      </div>
      <h3
        class="sc-VigVT jPSyOq"
      >
        Flights
      </h3>
      <p
        class="sc-jTzLTM entxYC"
      />
      <a
        class="sc-fjdhpX grVBQl"
        href="/trip/1/budget-category/1/budget-item/1/expense/1"
      >
        <div
          class="sc-jzJRlG eFxIKS"
        >
          <div
            class="sc-kAzzGY jMDNuo"
          >
            Apr 26
          </div>
          <div
            class="sc-chPdSV iSnHzR"
          >
            Edit
          </div>
        </div>
        <div
          class="sc-cSHVUG fqivQQ"
        >
          <h4
            class="sc-kgoBCf iCrCjs"
          >
            Southwest
          </h4>
          <div
            class="sc-kGXeez eIIwaK"
          >
            $594.62
          </div>
        </div>
        <p
          class="sc-kpOJdX guChom"
        />
      </a>
    </div>
  </div>
</DocumentFragment>
`;
